core/src/main/java/org/infinispan/marshall/core/internal/InternalMarshaller.java

```diff
@@ -0,0 +1,227 @@
+package org.infinispan.marshall.core.internal;
+
+import org.infinispan.commands.RemoteCommandsFactory;
+import org.infinispan.commons.io.ByteBuffer;
+import org.infinispan.commons.io.ByteBufferImpl;
+import org.infinispan.commons.io.ExposedByteArrayOutputStream;
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.marshall.BufferSizePredictor;
+import org.infinispan.commons.marshall.Externalizer;
+import org.infinispan.commons.marshall.MarshallableTypeHints;
+import org.infinispan.commons.marshall.NotSerializableException;
+import org.infinispan.commons.marshall.SerializeWith;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.factories.GlobalComponentRegistry;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.OutputStream;
+import java.io.Serializable;
+
+// TODO: If wrapped around GlobalMarshaller, there might not be need to implement StreamingMarshaller interface
+// If exposed directly, e.g. not wrapped by GlobalMarshaller, it'd need to implement StreamingMarshaller
+public final class InternalMarshaller implements StreamingMarshaller {
+
+   private static final Log log = LogFactory.getLog(InternalMarshaller.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   final MarshallableTypeHints marshallableTypeHints = new MarshallableTypeHints();
+
+   final Encoding<BytesObjectOutput, BytesObjectInput> enc = new BytesEncoding();
+   final InternalExternalizerTable externalizers;
+
+   final StreamingMarshaller external;
+
+   public InternalMarshaller(GlobalComponentRegistry gcr, RemoteCommandsFactory cmdFactory) {
+      this.externalizers = new InternalExternalizerTable(enc, gcr, cmdFactory);
+      //this.external = new ExternalJavaMarshaller();
+      this.external = new ExternalJBossMarshaller(externalizers, gcr.getGlobalConfiguration());
```

Comment:
The external marshaller should definitely be configurable (unlike the internal one).